client/branded/src/search-ui/input/experimental/CodeMirrorQueryInputWrapper.tsx

```diff
@@ -420,31 +430,48 @@ export const CodeMirrorQueryInputWrapper = forwardRef<Editor, PropsWithChildren<
             }
         }, [setMode])
 
-        const { ref: spacerRef, height: spacerHeight } = useResizeObserver({
-            ref: focusContainerRef,
-        })
-
         return (
-            <div className={styles.container}>
-                {/* eslint-disable-next-line react/forbid-dom-props */}
-                <div className={styles.spacer} style={{ height: `${spacerHeight}px` }} />
-                <div className={styles.root}>
-                    <div ref={spacerRef} className={styles.focusContainer}>
-                        <div className={classNames(styles.modeSection, !!mode && styles.active)}>
-                            <Tooltip content="Recent searches">
-                                <Button variant="icon" onClick={toggleHistoryMode} aria-label="Open search history">
-                                    <Icon svgPath={mdiClockOutline} aria-hidden="true" />
-                                </Button>
-                            </Tooltip>
-                            {mode && <span className="ml-1">{mode}:</span>}
-                        </div>
-                        <div ref={editorContainerRef} className={styles.input} />
-                        {!mode && children}
-                    </div>
-                    <div ref={setSuggestionsContainer} className={styles.suggestions} />
+            <div
+                className={classNames(styles.container, className, {
+                    [styles.containerCompact]: visualMode === QueryInputVisualMode.Compact,
+                })}
+            >
+                <div className={styles.focusContainer}>
+                    <SearchModeSwitcher mode={mode} onModeChange={toggleHistoryMode} />
+                    <div ref={editorContainerRef} className={styles.input} />
+                    {!mode && children}
                 </div>
+                <div ref={setSuggestionsContainer} className={styles.suggestions} />
                 <Shortcut ordered={['/']} onMatch={focus} />
             </div>
         )
     }
 )
+
+interface SearchModeSwitcherProps {
+    mode: string | null
+    className?: string
+    onModeChange: () => void
+}
+
+const SearchModeSwitcher: FC<SearchModeSwitcherProps> = props => {
+    const { mode, className, onModeChange } = props
+
+    return (
+        <div className={classNames(className, styles.mode, !!mode && styles.modeActive)}>
+            <Tooltip content="Recent searches">
+                <Button variant="icon" aria-label="Open search history" onClick={onModeChange}>
+                    <Icon svgPath={mdiClockOutline} aria-hidden="true" />
+                </Button>
+            </Tooltip>
+            {mode && <span className="ml-1">{mode}:</span>}
+        </div>
+    )
+}
+
+function useMutableValue<T>(value: T): MutableRefObject<T> {
+    const valueRef = useRef(value)
+    valueRef.current = value
```

Comment:
This needs to be 

```suggestion
    useEffect(() => {
      valueRef.current = value
    }, [value])
```

Refs are not supposed to be read or written to during render (as I recently learned).

Reply:
@fkling just our of curiosity, what may happen if we update value during render? I thought that refs are just plain JS objects so they retain their values between renders. 

I know that documentations says that we shouldn't do this https://beta.reactjs.org/learn/referencing-values-with-refs#best-practices-for-refs

but I just don't see how it would hurt in this particular case with callbacks.

Reply:
The update might never be committed and then the ref could potentially refer to an incorrect value, or something like that?
Regardless, I'd follow to docs and be consistent. We cannot except everyone to understand the finer nuances of React.

Reply:
Agree, I will apply you suggestions with useEffect

>The update might never be committed and then the ref could potentially refer to an incorrect value, or something like that?

To be honest, I still don't understand, even if something has been interrupted and never committed this won't happen in a middle of the component. So the component itself should be in  inconsistent state.  I see some cases with closure when we read and write in rendering but in this case it always read in user input based callbacks, where when we read everything is settled

Reply:
If we really want to align with approach with user effect I suggest to talk about on our FE crew because our codebase is full of mutable write operation in rendering, if you have some specific examples it would be even cooler